Runner behavior refactoring

s1mplex opened the pull request:
My company's preparing to run a small and unusual Hackathon and our vision is to have Atom as a default text editor with a few plugins that will help us with hosting it.

We decided to use Atom-Script as a runner, but it lacks some functionality so I decided to do some refactoring so it's implemented easier.

* **Custom input source:** I made it possible to customize an input provider that accepts Node's stdin Stream.
*  **Custom formatting options:** I don't always want to show the output in an Atom view For example for our event we're planning to make a web request with the output and receive feedback about the solution. I believe some people'd also like to write to a file instead to the view. I used the subscription model following the conventions from [this post](http://blog.atom.io/2014/09/16/new-event-subscription-api.html)

Please let me know what you think about this!

s1mplex commented:
@rgbkrk @erran Any chance you could take a look at it?

rgbkrk commented:
Thanks for the ping and major thanks for tackling a refactor. Checking it out now!

rgbkrk commented on lib/runner.coffee:
```diff
@@ -0,0 +1,89 @@
+{Emitter, BufferedProcess} = require 'atom'
```
If you're going to these lengths, how would you feel about using `child_process.spawn`?

s1mplex replied:
Good idea! I'll rewrite it tomorrow using Node's API

s1mplex replied:
@rgbkrk I just did a spike on that and I feel it'll either complicate things or make them less extensible.

Lets say I only use the `child_process` API. Then when I want to have more than one Formatter, Runner should either 

* be aware that it can call multiple callback functions (eg. one `stdoutFunc` per each Formatter)
* implement custom event system (Emitter does it for me)

I'd personally stick with the current implementation unless you had a bigger idea in mind (I have rather zero-ish experience in both Atom and Node APIs)

rgbkrk replied:
Ah, that's ok then. I'll review as is and take it for a spin.

rgbkrk commented on lib/view-formatter.coffee:
```diff
@@ -0,0 +1,17 @@
+    @subscriptions.add @runner.onDidExit (ev) =>
+      @view.setHeaderAndShowExecutionTime ev.returnCode, ev.executionTime
+    @subscriptions.add @runner.onDidNotRun (ev) =>
+      @view.showUnableToRunError ev.command
```
This is looking great.

s1mplex commented:
@rgbkrk Hopefully you haven't had a look at it yet because I've found a bug ;) it's now fixed and I added some tests to the Runner class. Let me know if you feel like I'm missing anything in this PR!

rgbkrk commented:
Uh oh! Were you still looking to include this? Sorry I had not come back to it.

s1mplex commented:
@rgbkrk I decided I need some time to make a better refactoring with more tests included and with exposed services. I'll make a new PR, just give me some time! ;)
